### PR TITLE
add logs and allow configurabe log level and mode

### PR DESCRIPTION
**Pull request by shan4993**

### CODE REVIEW

File: pkg/retry/retry.go [hunk 1]
```diff
@@ -61,7 +62,12 @@ func (s *Server) handleRetry() http.Handler {
 			return
 		}
 
-		for cursor != "" {
+		logger.Infow("retrieved last checkpoint", "checkpoint", cursor)
+
+		// for each run track the total number of events retrieve and failure events
+		totalEventCount, failedEventCount := 0, 0
+
+		for ok := true; ok; ok = (cursor != "") {
```
Comment (shan4993) [inline pkg/retry/retry.go:70]: @sethvargo i brought this back because we need the loop to run at least once. The first time this service is invoked, the checkpoint will be empty. I'm attempting to replicate a do while loop in go. If you agree, is this the correct way?

Comment (verbanicm) [inline pkg/retry/retry.go:70]: This doesnt seem right? Why not just cursor != ""

Comment (verbanicm) [reply pkg/retry/retry.go:70]: Ill defer to your comment for Seth for the idiomatic approach.

Comment (sethvargo) [reply pkg/retry/retry.go:70]: Yea this is common. The other option is to `break` under a condition, but this is fine.

File: pkg/retry/retry.go [hunk 2]
```diff
@@ -61,7 +62,12 @@ func (s *Server) handleRetry() http.Handler {
 			return
 		}
 
-		for cursor != "" {
+		logger.Infow("retrieved last checkpoint", "checkpoint", cursor)
+
+		// for each run track the total number of events retrieve and failure events
+		totalEventCount, failedEventCount := 0, 0
```
Comment (sethvargo) [inline pkg/retry/retry.go:68]: Nit, it's more idiomatic to do:

```go
var totalEventCount int
var failedEventCount int
```

File: pkg/retry/retry.go [hunk 3]
```diff
@@ -103,7 +119,8 @@ func (s *Server) handleRetry() http.Handler {
 			}
 		}
 
-		logger.Infow("successful", "code", http.StatusAccepted, "body", acceptedMessage)
+		logger.Infow("successful", "code", http.StatusAccepted, "body", acceptedMessage, "totalEventCount", totalEventCount,
```
Comment (sethvargo) [inline pkg/retry/retry.go:122]: Nit: when the log gets long, split it to KV pairs:

```suggestion
		logger.Infow("successful",
			"code", http.StatusAccepted,
			"body", acceptedMessage,
			"totalEventCount", totalEventCount,
```

File: pkg/retry/retry.go [hunk 4]
```diff
@@ -83,18 +94,23 @@ func (s *Server) handleRetry() http.Handler {
...
+
+				// append to the total failure events counter
+				failedEventCount += 1
 			}
 
-			// periodic update to checkpoint
-			if err := s.datastore.WriteCheckpointID(ctx, s.checkpointTableID, "checkpointID", time.Now().String()); err != nil {
+			logger.Infow("write latest checkpoint", "checkpoint", cursor)
+
+			if err := s.datastore.WriteCheckpointID(ctx, s.checkpointTableID, cursor, time.Now().String()); err != nil {
```
Comment (sethvargo) [inline pkg/retry/retry.go:113]: 1. Please grab a copy of `time.Now()` at the very start of the invocation, otherwise you'll have races.

2. Please use `UTC` of every and all times.

3. Is `String` the correct function? I would expect `.Format()` in a specific format instead.

File: terraform/variables.tf [hunk 5]
```diff
@@ -231,7 +231,7 @@ variable "lock_ttl_clock_skew" {
 }
 
 variable "cloud_scheduler_deadline_duration" {
-  description = "The deadline for job attempts in seconds. If the request handler does not respond by this deadline then the request is cancelled and the attempt is marked as a DEADLINE_EXCEEDED failure. Defaults to 30 minutes (max)"
+  description = "The deadline for job attempts in seconds. If the request handler does not respond by this deadline then the request is cancelled and the attempt is marked as a DEADLINE_EXCEEDED failure. Default is 30 minutes (max)"
```
Comment (sethvargo) [inline terraform/variables.tf:234]: Why this change? I think this is no longer grammatically correct...

File: terraform/variables.tf [hunk 6]
```diff
@@ -274,3 +274,15 @@ variable "region" {
   type        = string
   default     = "us-central1"
 }
+
+variable "log_mode" {
+  description = "The mode for logging. Default log mode is production."
```
Comment (sethvargo) [inline terraform/variables.tf:279]:
```suggestion
  description = "The mode for logging. Defaults to 'production'."
```

File: terraform/variables.tf [hunk 7]
```diff
@@ -274,3 +274,15 @@ variable "region" {
...
+  type        = string
+  default     = "production"
+}
+
+variable "log_level" {
+  description = "The level for logging. Default log level is warning."
```
Comment (sethvargo) [inline terraform/variables.tf:285]:
```suggestion
  description = "The level for logging. Defaults to 'warning'."
```

File: terraform/variables.tf (hunk 5 above)
Comment (verbanicm) [reply terraform/variables.tf:234]: It previously has a typo `Default is to` so I suggested `Default is`, but I think you're right.

`Defaults to 30 minutes`

**Review by verbanicm — APPROVED** [review]

File: pkg/retry/retry.go [hunk 8]
```diff
@@ -37,41 +37,65 @@ const (
 // events.
 func (s *Server) handleRetry() http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		now := time.Now().UTC().Format(time.UnixDate)
```
Comment (sethvargo) [inline pkg/retry/retry.go:40]: Sorry, nit: I'd make this `time.Now().UTC()` so you have a real time object that you can use throughout.

Then when you need to put it in the query, call `.Format(time.UnixDate)`.

**Review by sethvargo — APPROVED** [review]
